### PR TITLE
Recover player from a broken state

### PR DESCRIPTION
**Pull request by AJenbo**
This allows for recovering broken players and corrects some stats based on calculations based on other values.

### CODE REVIEW

File: Source/pack.cpp [hunk 1]
```diff
@@ -289,45 +289,51 @@ void UnPackPlayer(const PlayerPack &packed, Player &player)
...
+
 	player._pMaxHPBase = SDL_SwapLE32(packed.pMaxHPBase);
+	player._pMaxHPBase = std::min(player._pMaxHPBase, player.calculateBaseLife());
 	player._pHPBase = SDL_SwapLE32(packed.pHPBase);
 	player._pHPBase = clamp<int32_t>(player._pHPBase, 0, player._pMaxHPBase);
 	player._pMaxHP = player._pMaxHPBase;
 	player._pHitPoints = player._pHPBase;
+
+	player._pMaxManaBase = SDL_SwapLE32(packed.pMaxManaBase);
+	player._pMaxManaBase = std::min(player._pMaxManaBase, player.calculateBaseMana());
```
Comment (AJenbo) [inline Source/pack.cpp:312]:
```suggestion
	player._pMaxHPBase = std::min<int32_t>(player._pMaxHPBase, player.calculateBaseLife());
	player._pHPBase = SDL_SwapLE32(packed.pHPBase);
	player._pHPBase = clamp<int32_t>(player._pHPBase, 0, player._pMaxHPBase);
	player._pMaxHP = player._pMaxHPBase;
	player._pHitPoints = player._pHPBase;

	player._pMaxManaBase = SDL_SwapLE32(packed.pMaxManaBase);
	player._pMaxManaBase = std::min<int32_t>(player._pMaxManaBase, player.calculateBaseMana());

```

**Review by StephenCWills — APPROVED** [review]

File: Source/pack.cpp [hunk 2]
```diff
@@ -289,45 +289,52 @@ void UnPackPlayer(const PlayerPack &packed, Player &player)
...
 	player.position.future = position;
 	player.setLevel(clamp<int8_t>(packed.plrlevel, 0, NUMLEVELS));
 
-	player._pClass = static_cast<HeroClass>(clamp<uint8_t>(packed.pClass, 0, enum_size<HeroClass>::value - 1));
-
 	ClrPlrPath(player);
 	player.destAction = ACTION_NONE;
 
 	CopyUtf8(player._pName, packed.pName, sizeof(player._pName));
+	if (strlen(player._pName) == 0) {
```
Comment (glebm) [inline Source/pack.cpp:326]: No need to calculate the length just to check if it's empty:

```suggestion
	if (player._pName[0] == '\0') {
```

**Review by glebm — APPROVED** [review]

**Comment by jstasiak** [comment]
I had a game today where this patch seems like it would've helped, this line specifically:

```
player._pMaxHPBase = std::min<int32_t>(player._pMaxHPBase, player.calculateBaseLife());
```

The character in question somehow got into a broken state and fails to join a game because

```
DEBUG: Remote player validation failed: ValidateFields(player._pMaxHPBase: 6208, player.calculateBaseLife(): 4928, player._pMaxHPBase <= player.calculateBaseLife())
```

Both client and server running version 1.5.1.

**Comment by StephenCWills** [comment]
I'm much more interested in how that could have happened. It seems you have 20 HP more than you're supposed to.

**Comment by AJenbo** [comment]
Can you post a save file?

**Comment by jstasiak** [comment]
Sadly the character got unbroken when a game was created with it and it died, it went back to the original HP and there's no save file with the broken state.

Before it broke the character had max 37 HP. In the broken state that went up to 100 HP as displayed in-game. Now it's back to 37 HP.


> I'm much more interested in how that could have happened. It seems you have 20 HP more than you're supposed to.

Yeah that's fair. The client process with the character in question crashed when it attempted to enter The Catacombs level 2 or 3 but it's unclear now if that was through a town portal or from another level. The character became broken then.

We hit this bug relatively quickly after starting to play the game so I'd say it's likely we'll hit it again, in which case I'll try to gather more information and provide a save file.

**Pull request closed without merging**